tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/build_soma.py

```diff
@@ -104,6 +109,17 @@ def build(args: CensusBuildArgs) -> int:
     return 0
 
 
+def prune_unused_datasets(assets_path: pathlib.Path, all_datasets: List[Dataset], used_datasets: List[Dataset]) -> None:
+    unused_datasets = [d for d in all_datasets if d.dataset_id not in set(d.dataset_id for d in used_datasets)]
```

Comment:
nit: use set.difference() or `-` for readability?

Reply:
I'm not clear how that simplifies things given that the Dataset object isn't hashable, and I need to end up with a list of them.

If I understand, you are proposing:
>     unused_datasets = [d for d in all_datasets if d.dataset_id in (set(d.dataset_id for d in all_datasets) - set(d.dataset_id for d in used_datasets))]

which seems verbose and not any clearer.

I guess I could allocate another data structure at some expense which is hashable, eg.,, which would look like:
```
    all_datasets_by_id = {d.dataset_id: d for d in all_datasets}
    unused_dataset_ids = {d.dataset_id for d in used_datasets}
    used_dataset_ids = set(all_datasets_by_id) - unused_dataset_ids
    assert all(all_datasets_by_id[id].dataset_total_cell_count == 0 for id in unused_datasets)
    assert all(all_datasets_by_id[id].dataset_total_cell_count > 0 for id in used_datasets)
    assert set(all_datasets_by_id[id].dataset_id for id in used_datasets).isdisjoint(
        set(all_datasets_by_id[id].dataset_id for d in unused_datasets)
    )

    for dataset_id in used_dataset_ids:
        os.remove(assets_path / all_datasets_by_id[dataset_id].dataset_h5ad_path)
```

Also doesn't seem like a win.

Can you clarify your intent?  I think I'm missing something.

Reply:
Indeed, dataset objs would need to be hashable for improve readability, so best as is!